Tests/MongoSwiftTests/DocumentTests.swift

```diff
@@ -372,4 +373,131 @@ final class DocumentTests: XCTestCase {
         expect(doc["a1"]).to(equal(arr1))
         expect(doc["a2"]).to(equal(arr2))
     }
+
+    // exclude Int64 value on 32-bit platforms
+    static let overwritables: Document = XCTestCase.is32Bit ?
+                                        ["double": 2.5, "int32": Int32(32), "bool": false, "decimal": Decimal128("1.2E+10")] :
+                                        ["double": 2.5, "int32": Int32(32), "int64": Int64(64), "bool": false, "decimal": Decimal128("1.2E+10")]
+
+    static let nonOverwritables: Document = ["string": "hello", "nil": nil, "doc": ["x": 1] as Document, "arr": [1, 2] as [Int], "oid": ObjectId()]
+
+    // test replacing `Overwritable` types with values of their own type
+    func testOverwritable() throws {
+        var doc = DocumentTests.overwritables
+
+        // overwrite int32 with int32
+        doc["int32"] = Int32(15)
+        expect(doc["int32"] as? Int).to(equal(15))
+
+        // overwrite int32 with int
+        doc["int32"] = 20
+        expect(doc["int32"] as? Int).to(equal(20))
+
+        doc["bool"] = true
+        doc["double"] = 3.0
+        doc["decimal"] = Decimal128("100")
+
+        if XCTestCase.is32Bit {
+            expect(doc).to(equal(["double": 3.0, "int32": 20, "bool": true, "decimal": Decimal128("100")]))
+            return
+        }
+
+        // overwrite int64 with int64
+        let bigInt64 = Int64(Int32.max) + 1
```

Comment:
Is there actually a problem working with `Int64` objects on a 32-bit platform? IIRC, the skip in `testIntEncodesAsInt32OrInt64()` was only used because we were doing arithmetic.

In this case, I wonder if you could just create `Int64` objects from `Int64.max` and overwrite it with `Int64.min`.

Reply:
ah, yeah you're right. in this case, I will switch to overwriting with Int64.min. 

However, for the next part, where I want to try replacing an `Int64` with an `Int`, I need an `Int` whose `bsonType` will be `Int64` (which will only be the case if `Int32(exactly:)` fails). I don't think an `Int` larger than `Int32.max` can exist on a 32 bit platform, right? but we can at least move the early return a bit later.

Reply:
> I don't think an Int larger than Int32.max can exist on a 32 bit platform, right?

Assuming `Int` is platform dependent, I'd agree.

> However, for the next part, where I want to try replacing an Int64 with an Int

Since libbson requires integers to be replaced with the same size, would it make sense that we should allow an Int64 BSON type to be replaced with any integer? If it's in the range of 32-bit values (e.g. result of decrementing a value in the 64-bit range), perhaps we should cast it up to a 64-bit value to still take advantage of in-place replacement in libbson. Obviously, the reverse case (`Int` in 32-bit range growing to 64-bit range) would require re-encoding and could not utilize in-place replacement.

I'm not sure how the server handles this (if it still does). With MMAPv1, there were optimizations for in-place replacements, but I don't think those apply to WT, since documents are always re-written.

Reply:
[it is](https://developer.apple.com/documentation/swift/int) - 
> On 32-bit platforms, Int is the same size as Int32, and on 64-bit platforms, Int is the same size as Int64.

I was thinking about that too. my only concern was that when reading the value out of the document, the user would then get an `Int64` when they expect an `Int32`. I can't think of any situations where that would be really catastrophic, but it may be unexpected. what do you think?

Reply:
Sorry for losing track of this conversation. I don't follow the concern you're raising.

IIRC, we were discussing allowing an Int64 to be overwritten by _any_ integer value. If an Int, which happens to be an Int32, or an explicit Int32 is passed, we can cast it up and widen it to an Int64 in order to do an in-place update in libbson.

I'm not sure how this would adversely affect a user. Even if they did want to assert that they had overwritten a large-than-32-bit value with a value within the 32-bit range, Swift should still allow them to downcast safely to an Int32 when accessing the field (I assume using `as Int32`), no?

Reply:
unfortunately no, consider the following snippet:
```swift
let doc: Document = ["a": Int64(100)]
print(doc["a"] as? Int32) // nil
print(doc["a"] as? Int64) // 100
```
the subscript operator has no knowledge of what type we're requesting, so it just returns a value of whatever Swift type corresponds to BSON type specified in the `bson_t`. 

we aren't allowed to cast `Int64` to `Int32` using `as`, so a user would have to do something like
```swift
guard let int64Value = doc["a"] as? Int64 else { /// ... }
let int32Value = Int32(int64Value)
```
so if the user needs to be able to extract the field specifically as an `Int32` this would be problematic.

Reply:
And it looks like we also can't do `doc["a"] as? Int` to return the Int64 as an Int, even if we're on a 64-bit platform. `DocumentIterator.BsonTypeMap` says that all BSON int32 and int64 types are converted to an `Int` and `Int64`, respectively.

I wonder if the existing behavior is problematic as well. Consider:

```swift
let a = Int(Int64.max)
let doc: Document = ["a": a]
print(a) // 9223372036854775807
print(doc["a"] as? Int) // nil
print(doc["a"] as? Int64) // Optional(9223372036854775807)
```

If users are accessing a BSON integer field, they must anticipate its range before they can cast it to a native type. And they can't rely on always casting to `Int64`, as a BSON int32 type will yield nil if cast to `Int64`.

To bring this back to the current PR, I think it's sensible to only allow overwriting (really just an optimization) when the exact BSON type is used. Separate from that, it seems like there's room for improvement with integer access.

Reply:
This can be addressed by [SWIFT-221](https://jira.mongodb.org/browse/SWIFT-221). Resolving.